### PR TITLE
Add context support to Atomic Blocks

### PR DESCRIPTION
**Pull request by mikejolley**
#2399 is a mammoth PR so I'm breaking out some fixes and changes to review separately.

This PR makes our Atomic Blocks context ready. The context support is backwards compatible with the old way of passing a product as a prop - that takes priority if used.

Changes:

- Refactored Atomic Blocks to not break when no product is given - this is important whilst context is loaded.
- Renamed `useProductDataContextContext` to `useProductDataContext`
- Renamed `useInnerBlockConfigurationContext` to `useInnerBlockLayoutContext` to better reflect its use.
- Added context handling to each Atomic Block
- Removed default "preview" data and instead injected it via the new context!

### How to test the changes in this Pull Request:

1. Using All Products Block, check there are no regressions in the layout edit view, in particular, check the preview data is shown correctly since this is now coming from context.
2. Check frontend view of All Products for regressions. Add to cart buttons etc.

That should cover it. Nothing else is using context except the layout edit preview data, so just testing all works as before is all thats needed.

### CODE REVIEW

File: assets/js/blocks/products/all-products/edit.js [hunk 1]
```diff
@@ -203,7 +207,11 @@ class Editor extends Component {
 					<div className="wc-block-grid has-1-columns">
 						<ul className="wc-block-grid__products">
 							<li className="wc-block-grid__product">
-								<InnerBlocks { ...InnerBlockProps } />
+								<ProductDataContextProvider
```
Comment (mikejolley) [inline assets/js/blocks/products/all-products/edit.js:210]: Preview data is passed to the context here.

File: assets/js/atomic/blocks/product/title/block.js [hunk 2]
```diff
@@ -3,30 +3,57 @@
...
 	headingLevel = 2,
 	productLink = true,
+	...props
 } ) => {
-	const { layoutStyleClassPrefix } = useInnerBlockConfigurationContext();
-	if ( ! product.name ) {
-		return null;
+	const productDataContext = { ...useProductDataContext(), ...props };
+	const { layoutStyleClassPrefix } = useInnerBlockLayoutContext();
+	const { product } = productDataContext;
```
Comment (Aljullu) [inline assets/js/atomic/blocks/product/title/block.js:31]: Personal opinion here, but these two lines were confusing me a bit:
```JS
const productDataContext = { ...useProductDataContext(), ...props };
const { product } = productDataContext;
```
I think this can be rewritten so the spread syntax is not necessary:
```JS
const productDataContext = useProductDataContext();
const { product } = productDataContext || props;
```
I know it's just personal taste, but proposing it in case you like it (it's affecting all other inner blocks). :slightly_smiling_face:

File: assets/js/atomic/blocks/product/image/block.js [hunk 3]
```diff
@@ -2,101 +2,129 @@
...
-		<div
-			className={ classnames(
-				className,
-				`${ layoutStyleClassPrefix }__product-image`
-			) }
-		>
+		<div className={ classnames( className, componentClass ) }>
 			{ productLink ? (
 				<a href={ product.permalink } rel="nofollow">
-					{ renderedSalesAndImage }
```
Comment (Aljullu) [inline assets/js/atomic/blocks/product/image/block.js:88]: Just curious, why not keeping `renderedSalesAndImage`? It saved some duplicated lines of code.

Comment (mikejolley) [reply assets/js/atomic/blocks/product/image/block.js:88]: I found it a little hard to follow the logic before, where now it's in two clear blocks.

File: assets/js/atomic/blocks/product/title/block.js (hunk 2 above)
Comment (mikejolley) [reply assets/js/atomic/blocks/product/title/block.js]: @Aljullu This breaks :) Context has value - we just want to take the product from it, or props, with props taking priority.

Comment (mikejolley) [reply assets/js/atomic/blocks/product/title/block.js]: This works

```
const productDataContext = useProductDataContext();
const product = props.product || productDataContext.product;
```

**Comment by mikejolley** [comment]
> Code looks good and is testing well, but I get a JS error when editing the All Products block in the editor (pressing the pencil icon):

Not happening here @Aljullu 

Made the other changes requested :)
